File: capacitor-plugins/capacitor-native/ios/Sources/AonsokuNativePlugin/Audio/NativeSourceResolver.swift
import Foundation

class NativeSourceResolver {
    let cacheDirectories: [URL]
    private var cachedCredentials: ServerCredentials?
    private var credentialsCacheTime: Date?
    private let credentialsTTL: TimeInterval = 30

    init() {
        var dirs: [URL] = []

        let appSupport = FileManager.default.urls(
            for: .applicationSupportDirectory, in: .userDomainMask
        ).first!
        dirs.append(
            appSupport
                .appendingPathComponent("Aonsoku", isDirectory: true)
                .appendingPathComponent("AudioCache", isDirectory: true)
        )

        if let documents = FileManager.default.urls(
            for: .documentDirectory, in: .userDomainMask
        ).first {
            dirs.append(
                documents.appendingPathComponent("AudioCache", isDirectory: true)
            )
        }

        self.cacheDirectories = dirs
    }

    func invalidateCredentialsCache() {
        cachedCredentials = nil
        credentialsCacheTime = nil
    }

    private func getCredentials() -> ServerCredentials? {
        if let cached = cachedCredentials,
           let cacheTime = credentialsCacheTime,
           Date().timeIntervalSince(cacheTime) < credentialsTTL {
            return cached
        }
        let creds = KeychainManager.retrieve()
        cachedCredentials = creds
        credentialsCacheTime = Date()
        return creds
    }

    func resolveSource(for song: QueueSong) -> (url: URL, kind: String)? {
        if let cachedUri = song.cachedFileUri, !cachedUri.isEmpty {
            let fileUrl = fileURL(from: cachedUri)
            if FileManager.default.fileExists(atPath: fileUrl.path) {
                return (fileUrl, "native-file")
            }
        }

        let cacheId = cacheId(for: song.id)
        let extensions = ["mp3", "flac", "m4a", "aac", "ogg", "opus", "wav", "audio"]
        for directory in cacheDirectories {
            guard FileManager.default.fileExists(atPath: directory.path) else {
                NativeLogger.shared.debug("SourceResolver: directory not found: \(directory.path)", source: "Audio")
                continue
            }
            for ext in extensions {
                let fileUrl = directory.appendingPathComponent("\(cacheId).\(ext)")
                if FileManager.default.fileExists(atPath: fileUrl.path) {
                    NativeLogger.shared.info("SourceResolver: found cached file for \(song.id) at \(fileUrl.lastPathComponent)", source: "Audio")
                    return (fileUrl, "native-file")
                }
            }
        }
        NativeLogger.shared.debug("SourceResolver: no cache hit for \(song.id), cacheId=\(cacheId)", source: "Audio")

        // Extract songId from aonsoku-media://stream?id=... URLs if present
        let effectiveSongId: String
        if song.streamUrl.hasPrefix("aonsoku-media://stream"),
           let components = URLComponents(string: song.streamUrl),
           let queryItems = components.queryItems,
           let idItem = queryItems.first(where: { $0.name == "id" }),
           let idValue = idItem.value, !idValue.isEmpty {
            effectiveSongId = idValue
        } else {
            effectiveSongId = song.id
        }

        if let url = buildAuthenticatedStreamUrl(songId: effectiveSongId) {
            return (url, "stream")
        }

        // Fallback: if streamUrl itself is a valid absolute URL, try to use it directly
        if let streamUrl = URL(string: song.streamUrl),
           streamUrl.scheme != nil,
           streamUrl.host != nil {
            return (streamUrl, "stream")
        }

        return nil
    }

    func buildStreamUrl(songId: String) -> String? {
        buildAuthenticatedStreamUrl(songId: songId)?.absoluteString
    }

    private func buildAuthenticatedStreamUrl(songId: String) -> URL? {
        guard let credentials = getCredentials() else { return nil }

        var params = SubsonicAuthBuilder.buildQueryParams(
            username: credentials.username,
            password: credentials.password,
            authType: credentials.authType,
            protocolVersion: credentials.protocolVersion
        )
        params["id"] = songId
        // Must be "true" for native AVPlayer: without Content-Length, AVPlayer cannot
        // issue HTTP Range requests, causing buffer stalls and seek failures.
        // (Web uses "false" to work around browser-specific streaming quirks.)
        params["estimateContentLength"] = "true"

        let baseString = "\(credentials.serverUrl)/rest/stream"
        guard var components = URLComponents(string: baseString) else { return nil }
        components.queryItems = params.map { URLQueryItem(name: $0.key, value: $0.value) }
        return components.url
    }

    private func cacheId(for songId: String) -> String {
        AudioCacheUtils.cacheId(for: songId)
    }

    private func fileURL(from uri: String) -> URL {
        if let url = URL(string: uri), url.isFileURL {
            return url
        }

        return URL(fileURLWithPath: uri)
    }
}

File: capacitor-plugins/capacitor-native/android/src/main/java/github/realtvop/aonsoku/plugins/audio/NativeSourceResolver.kt
package github.realtvop.aonsoku.plugins.audio

import android.content.Context
import android.net.Uri
import github.realtvop.aonsoku.plugins.bridge.AndroidCredentialStore
import github.realtvop.aonsoku.plugins.bridge.SubsonicAuthBuilder
import github.realtvop.aonsoku.plugins.debug.NativeLogger
import java.io.File
import java.net.URLEncoder

class NativeSourceResolver(
    private val context: Context,
    private val credentialStore: AndroidCredentialStore = AndroidCredentialStore(context)
) {
    private val cacheDirectories = listOf(
        AudioCacheUtils.getCacheDirectory(context),
        AudioCacheUtils.getSecondaryCacheDirectory(context)
    )

    fun resolveSource(song: QueueSong): Pair<String, String>? {
        // 1. If song.cachedFileUri is provided, check if it exists on disk
        if (!song.cachedFileUri.isNullOrEmpty()) {
            val path = if (song.cachedFileUri.startsWith("file://")) {
                song.cachedFileUri.substring(7)
            } else {
                song.cachedFileUri
            }
            val file = File(path)
            if (file.exists()) {
                return Pair("file://${file.absolutePath}", "native-file")
            }
            NativeLogger.warn("cachedFileUri does not exist on disk: $path", "source-resolver")
        }

        // 2. Check if the file is in local cache directories
        val cacheId = AudioCacheUtils.cacheId(song.id)
        val extensions = listOf("mp3", "flac", "m4a", "aac", "ogg", "opus", "wav", "audio")
        for (directory in cacheDirectories) {
            if (directory.exists() && directory.isDirectory) {
                for (ext in extensions) {
                    val file = File(directory, "$cacheId.$ext")
                    if (file.exists()) {
                        return Pair("file://${file.absolutePath}", "native-file")
                    }
                }
            }
        }

        // 3. Direct HTTP/HTTPS URL → pass through unchanged (radio, direct stream)
        if (song.streamUrl.startsWith("http://") || song.streamUrl.startsWith("https://")) {
            return Pair(song.streamUrl, "stream")
        }

        // 4. aonsoku-media://stream → resolve with credentials
        if (song.streamUrl.startsWith("aonsoku-media://stream")) {
            return resolveAonsokuStreamUrl(song)
        }

        NativeLogger.warn("Cannot resolve source: unknown streamUrl scheme for song ${song.id}", "source-resolver")
        return null
    }

    fun resolveCoverArtUrl(coverArtId: String, size: Int = 300): String? {
        val credentials = credentialStore.retrieve()
        if (credentials == null) {
            NativeLogger.warn("Cannot resolve cover art: no credentials", "source-resolver")
            return null
        }

        val cleanBaseUrl = credentials.serverUrl.trimEnd('/')
        val authParams = SubsonicAuthBuilder.buildQueryParams(
            username = credentials.username,
            password = credentials.password,
            authType = credentials.authType,
            protocolVersion = credentials.protocolVersion
        )
        val params = authParams + mapOf("id" to coverArtId, "size" to size.toString())
        val queryString = params.entries.joinToString("&") { (k, v) ->
            "${URLEncoder.encode(k, "UTF-8")}=${URLEncoder.encode(v, "UTF-8")}"
        }
        return "$cleanBaseUrl/rest/getCoverArt?$queryString"
    }

    private fun resolveAonsokuStreamUrl(song: QueueSong): Pair<String, String>? {
        val uri = Uri.parse(song.streamUrl)
        val songId = uri.getQueryParameter("id") ?: song.id
        val maxBitRate = uri.getQueryParameter("maxBitRate")
        val format = uri.getQueryParameter("format")

        val credentials = credentialStore.retrieve()
        if (credentials == null) {
            NativeLogger.error("Cannot resolve aonsoku-media://stream: no credentials for song ${song.id}", "source-resolver")
            return null
        }

        val cleanBaseUrl = credentials.serverUrl.trimEnd('/')
        val authParams = SubsonicAuthBuilder.buildQueryParams(
            username = credentials.username,
            password = credentials.password,
            authType = credentials.authType,
            protocolVersion = credentials.protocolVersion
        )
        val extraParams = mutableMapOf(
            "id" to songId,
            "estimateContentLength" to "true"
        )
        if (maxBitRate != null) extraParams["maxBitRate"] = maxBitRate
        if (format != null) extraParams["format"] = format

        val allParams = authParams + extraParams
        val queryString = allParams.entries.joinToString("&") { (k, v) ->
            "${URLEncoder.encode(k, "UTF-8")}=${URLEncoder.encode(v, "UTF-8")}"
        }
        val resolvedUrl = "$cleanBaseUrl/rest/stream?$queryString"
        NativeLogger.debug("Resolved aonsoku-media://stream -> $resolvedUrl", "source-resolver")
        return Pair(resolvedUrl, "stream")
    }
}
